src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php

```diff
@@ -52,6 +52,16 @@ protected function processValue($value, $isRoot = false)
             if (!$v instanceof Reference) {
                 throw new InvalidArgumentException(sprintf('Invalid definition for service "%s": an array of references is expected as first argument when the "container.service_locator" tag is set, "%s" found for key "%s".', $this->currentId, \is_object($v) ? \get_class($v) : \gettype($v), $k));
             }
+
+            if (\is_int($k)) {
+                unset($arguments[0][$k]);
+
+                $k = (string) $v;
+
+                if (isset($arguments[0][$k])) {
+                    throw new InvalidArgumentException(sprintf('Invalid definition for service "%s": service key "%s" cannot be inherited from the service definition because the key already exists.', $this->currentId, $k));
```

Comment:
Per my comments above, I'm not sure about this. It should be removed to me.

Reply:
(but please add tests showing that the later wins over the former if you can, that'll help prove who is right here :) )

Reply:
Fine by me if all differences have been set aside. I've checked how the different configuration formats handled duplicate declarations and they don't seem work the same at all, so there doesn't seem to be a best way.

Reply:
Can you elaborate? What's the difference you spotted?
If we want to keep the exception (I'm fine with it also), I'd suggest a message like:
`Invalid service locator definition for "%s": key "%s" is defined twice.', $this->currentId, $k));

Reply:
Here's my examples:
```
// PHP - Result: service_b
    ->setArguments([[
        'service_a' => new Reference('service_a'),
        'service_a' => new Reference('service_b'),
    ]])

// YAML - Result: service_a
        arguments:
            -
                service_a: '@service_a'
                service_a: '@service_b'

// XML - Result: service_b
      <argument type="collection">
        <argument type="service" id="service_a" key="service_a" />
        <argument type="service" id="service_b" key="service_a" />
      </argument>
```

Not that how PHP handles duplicate array keys matters a lot, just thought it might give a direction for the right solution.